mne/beamformer/_ap.py

```diff
@@ -0,0 +1,765 @@
+            logger.info('Done (optimally)')
+            break
+        if itr == attr_dict['max_iter']:
+            logger.info('Done (max iteration)')
+
+    return s_ap_2, oris, sub_g_proj
+
+
+def _calculate_free_alternating_projections(data_arr, gain,
+                                            nsources, max_iter,
+                                            force_no_rep):
+    """Calculate free-orientation alternating projection.
+
+    Parameters
+    ----------
+    data_arr : array, shape (nchannels, times)
+        Filtered evoked data array.
+    gain : array, shape (nchannels, ndipoles)
+        Gain array.
+    nsources : int
+        The number of dipoles to estimate.
+    max_iter : int
+        Maximal iteration number of AP.
+
+    Returns
+    -------
+    ap_temp_tuple : tuple
+        See: _free_phase2.
+
+    """
+    print('calculating free-orientation alternating projection')
+    ndipoles = int(gain.shape[1] / 3)
+    attr_dict = {
+        'ndipoles': ndipoles,
+        'nsources': nsources,
+        'max_iter': max_iter
+    }
+
+    data_cov = _produce_data_cov(data_arr, attr_dict)
+    # ######################################
+    # 1st Phase
+    # (a) Initialization: search the 1st source location over the entire
+    # dipoles topographies space
+    # ######################################
+
+    print(' 1st phase : ')
+    ap_temp_tuple = _free_phase1a(attr_dict, gain, data_cov)
+    # ap_temp_tuple = (s_ap, oris, sub_g_proj)
+
+    # ######################################
+    # (b) Now, add one source at a time
+    # ######################################
+
+    ap_temp_tuple = _free_phase1b(attr_dict, gain, data_cov,
+                                  ap_temp_tuple, force_no_rep)
+    # ap_temp_tuple = (s_ap, oris, sub_g_proj)
+    print('current s_ap = {}'.format(ap_temp_tuple[0]))
+
+    # #####################################
+    # 2nd phase
+    # #####################################
+
+    ap_temp_tuple = _free_phase2(ap_temp_tuple, attr_dict,
+                                 data_cov, gain, force_no_rep)
+
+    return ap_temp_tuple
+
+
+def _free_ori_ap(wh_data, gain, nsources,
+                 forward, max_iter, force_no_rep):
+    """Branch of calculations dedicated to freely oriented dipoles."""
+    sol_tuple = \
+        _calculate_free_alternating_projections(wh_data, gain,
+                                                nsources, max_iter,
+                                                force_no_rep)
+    # sol_tuple = active_idx, active_orientations, active_idx_gain
+
+    sol = lstsq(sol_tuple[2], wh_data, rcond=None)[0]
+
+    gain_fwd = forward['sol']['data'].copy()
+    gain_fwd.shape = (gain_fwd.shape[0], -1, 3)
+    gain_active = gain_fwd[:, sol_tuple[0]]
+    gain_dip = (sol_tuple[1] * gain_active).sum(-1)
+    idx = np.array(sol_tuple[0])
+    active_set = np.array(
+        [[3 * idx, 3 * idx + 1, 3 * idx + 2]]
+    ).T.ravel()
```

Comment:
I thought I could simplify this, but seeing it I'm not sure if it's better; feel free to reject this suggestion if you think the existing way is clearer

```suggestion
    rep = (3,) + (1,) * idx.ndim
    active_set = (np.tile(idx, rep) * 3 + np.arange(3).reshape(rep)).T.ravel()
```

Reply:
I'll leave it as it is